v0.13.0: refiner/reviewer claude-code path is unrunnable — agent told to write .ferry/cc-output.json but Write tool not in allowlist

## Summary

The refiner and reviewer roles run on the claude-code path with `ROLE_ACCESS = 'read-only'`, which restricts `--allowedTools` to `READ_ONLY_NATIVE_TOOLS = ["Read", "Glob", "Grep"]`. However, `outcomePromptSuffix(role)` instructs **every** role — including the read-only ones — to:

> "As your LAST action, write your result as a single JSON object to `.ferry/cc-output.json`, then stop."

Since `Write` is not in the allowlist and there is no narrow `Write(.ferry/cc-output.json)` grant anywhere in cc-prepare, every refiner/reviewer claude-code run terminates with `type: result, subtype: success, is_error: false, permission_denials_count: ≥1` — but `.ferry/cc-output.json` is never written. `ferry-cc-apply` then fails with:

```
ferry-cc-apply: failed to read .ferry/cc-output.json: ENOENT: no such file or directory, …
Error: Process completed with exit code 1.
```

The `claude-code-action` exit code is `0` (the agent "succeeded" from its own perspective — it ran to completion within budget), so the failure surfaces only at the cc-apply step, with the underlying cause buried in the agent's session log.

## Reproduction

1. Fix issues #352 (bundled prompts ENOENT) and #353 (`id-token: write`).
2. Trigger a `ferry-refine` dispatch routed to claude-code on a ticket with at least one untouched sub-task slot.
3. **Run claude-code-action** ✅ — observe `permission_denials_count: 3` (or similar non-zero) in the `success` result block.
4. **Apply claude-code output** ❌ — `failed to read .ferry/cc-output.json: ENOENT`.

Observed in run: https://github.com/big-emotion/ethniafrica/actions/runs/26192830025/job/77065346527 (refiner, 17 turns, $0.26, 3 permission denials, output never written).

## Root cause

`src/lib/claude-code/job-permissions.ts` + `src/lib/cc-wrappers/native-tools.ts` (compiled into `cc-prepare-action.js`):

```js
var READ_ONLY_NATIVE_TOOLS = ["Read", "Glob", "Grep"];
var READ_WRITE_NATIVE_TOOLS = ["Bash", "Read", "Write", "Edit", "Glob", "Grep"];
var ROLE_ACCESS = { refiner: "read-only", reviewer: "read-only", developer: "read-write", iterator: "read-write" };

function nativeToolsForRole(role) {
  return ROLE_ACCESS[role] === "read-only" ? [...READ_ONLY_NATIVE_TOOLS] : [...READ_WRITE_NATIVE_TOOLS];
}

function buildClaudeArgs(...) {
  const allowedTools = [...nativeToolsForRole(input.role), ...mcpToolAllowlist(servers)];
  // …
  return ["--allowedTools", allowedTools.join(","), …];
}
```

And `src/lib/claude-code/output-artifact.ts`:

```js
var CC_OUTPUT_ARTIFACT_PATH = ".ferry/cc-output.json";

function outcomePromptSuffix(role) {
  const header = `\n\n---\nFINAL OUTPUT (claude-code path): … As your LAST action, write your result as a single JSON object to \`${CC_OUTPUT_ARTIFACT_PATH}\`, then stop.`;
  if (role === "reviewer") return `${header} Shape: ${REVIEWER_SHAPE}`;
  if (role === "refiner")  return `${header} Shape: ${REFINER_SHAPE} …`;
  // …
}
```

The contract carrier (`.ferry/cc-output.json`) requires `Write`, but read-only roles cannot use it. This is unrecoverable without a narrow grant.

## Expected behavior

Every role on the claude-code path can write its result to `.ferry/cc-output.json`, regardless of `ROLE_ACCESS`. Read-only roles should remain otherwise unable to mutate the working tree.

## Proposed fix — pick one

**Option 1 (recommended): narrow `Write(.ferry/cc-output.json)` grant for all roles.**

Claude CLI's `--allowedTools` accepts path-scoped rules like `Write(path)`. In `buildClaudeArgs`:

```js
const allowedTools = [
  ...nativeToolsForRole(input.role),
  `Write(${CC_OUTPUT_ARTIFACT_PATH})`,
  ...mcpToolAllowlist(servers),
];
```

This grants exactly the file the contract requires, nothing more. Read-only roles remain blocked from arbitrary writes.

**Option 2:** switch read-only roles to read-write but layer a `--disallowedTools` restriction.

Add `Write`, `Edit`, `Bash` to refiner/reviewer's allowed list, then disallow `Write(*)` and `Edit(*)` except `Write(.ferry/cc-output.json)`. More flags, same outcome — option 1 is simpler.

**Option 3:** bypass the file artifact entirely for read-only roles.

Use `claude-code-action`'s `structured_output` JSON-schema output channel (`--json-schema` in `claude_args` → `outputs.structured_output`). The agent emits its JSON in its final assistant message via a schema-validated channel, and `ferry-cc-apply` reads `${{ steps.cc-run.outputs.structured_output }}` instead of disk. Bigger architectural change but eliminates the read-write coupling for refiner/reviewer.

## Recommended fix path

Ship option 1 in v0.13.1 (single-line diff in `buildClaudeArgs`). Track option 3 as the v0.15.0 simplification once consumers are unblocked.

## Consumer workaround (until fix lands)

Insert a workflow step between **Prepare claude-code job** and **Run claude-code-action** that augments `claude_args` to append `,Write(.ferry/cc-output.json)` to the `--allowedTools` value. Pseudo-shell:

```yaml
- name: Patch claude_args to allow cc-output write
  id: patch-args
  run: |
    args=$(echo '${{ steps.cc-prepare.outputs.claude_args }}' \
      | jq -c '. as $a
        | (index("--allowedTools")) as $i
        | $a | .[$i+1] |= . + ",Write(.ferry/cc-output.json)"')
    echo "args=$args" >> "$GITHUB_OUTPUT"

- name: Run claude-code-action
  uses: anthropics/claude-code-action@…
  with:
    claude_code_oauth_token: ${{ secrets.CLAUDE_CODE_OAUTH_TOKEN }}
    prompt: ${{ steps.cc-prepare.outputs.prompt }}
    claude_args: ${{ steps.patch-args.outputs.args }}
```

Untested in the form above — the JSON-shape transformation may need a small tweak depending on how cc-prepare serializes the array.

## Impact

- **Severity:** P0 for refiner + reviewer on the claude-code path. They never produce output even though **Run claude-code-action** reports success.
- **Wasted spend:** each failed refiner run costs ~$0.25 + 17 turns of Sonnet — multiplied across consumers and tickets, this is meaningful before anyone realizes the failure mode.
- **Debug difficulty:** **Run claude-code-action** reports success, **Apply claude-code output** reports ENOENT, and the connection between the two requires either reading the session log (not currently surfaced to consumers) or counting `permission_denials_count` in the inline output.

## Related

- ADR-0006 §2 — claude-code execution path
- Companion issue #352 — bundled prompts ENOENT
- Companion issue #353 — missing `id-token: write` in templates

--- a/.ferry/cc-prepare-action.js
+++ b/.ferry/cc-prepare-action.js
@@ -6658,35 +6658,6 @@ function assertToolPolicyEnforcesNoAutoMerge(policy) {
   }
 }
 
-// src/lib/claude-code/claude-args.ts
-function buildClaudeArgs(input) {
-  const servers = input.mcpServers ?? [];
-  const allowedTools = [...nativeToolsForRole(input.role), ...mcpToolAllowlist(servers)];
-  const disallowedTools = [...NO_AUTO_MERGE_DENY];
-  assertToolPolicyEnforcesNoAutoMerge({ allowedTools, disallowedTools });
-  const args = [
-    "--append-system-prompt",
-    input.system,
-    "--allowedTools",
-    allowedTools.join(","),
-    "--disallowedTools",
-    disallowedTools.join(",")
-  ];
-  if (servers.length > 0) {
-    args.push("--mcp-config", JSON.stringify(toClaudeCodeMcpConfig(servers)));
-  }
-  if (input.maxTurns !== void 0) {
-    if (!Number.isInteger(input.maxTurns) || input.maxTurns <= 0) {
-      throw new Error(`max-turns must be a positive integer, got ${input.maxTurns}`);
-    }
-    args.push("--max-turns", String(input.maxTurns));
-  }
-  if (input.model !== void 0 && input.model.trim().length > 0) {
-    args.push("--model", input.model);
-  }
-  return args;
-}
-
 // src/lib/claude-code/output-artifact.ts
 var CC_OUTPUT_ARTIFACT_PATH = ".ferry/cc-output.json";
 var DONE_OUTCOMES = ["implemented", "already_satisfied", "blocked"];
@@ -6842,6 +6813,39 @@ FINAL OUTPUT (claude-code path): the \`done\`/\`finish_review\` tools are not av
   }
   const access = ROLE_ACCESS[role];
   return `${header} Commit and push your work with \`git\` first (there is no \`commit_progress\` tool on this path; access=${access}). Then write: ${DEV_ITER_SHAPE}`;
+}
+
+// src/lib/claude-code/claude-args.ts
+function buildClaudeArgs(input) {
+  const servers = input.mcpServers ?? [];
+  const allowedTools = [
+    ...nativeToolsForRole(input.role),
+    `Write(${CC_OUTPUT_ARTIFACT_PATH})`,
+    ...mcpToolAllowlist(servers)
+  ];
+  const disallowedTools = [...NO_AUTO_MERGE_DENY];
+  assertToolPolicyEnforcesNoAutoMerge({ allowedTools, disallowedTools });
+  const args = [
+    "--append-system-prompt",
+    input.system,
+    "--allowedTools",
+    allowedTools.join(","),
+    "--disallowedTools",
+    disallowedTools.join(",")
+  ];
+  if (servers.length > 0) {
+    args.push("--mcp-config", JSON.stringify(toClaudeCodeMcpConfig(servers)));
+  }
+  if (input.maxTurns !== void 0) {
+    if (!Number.isInteger(input.maxTurns) || input.maxTurns <= 0) {
+      throw new Error(`max-turns must be a positive integer, got ${input.maxTurns}`);
+    }
+    args.push("--max-turns", String(input.maxTurns));
+  }
+  if (input.model !== void 0 && input.model.trim().length > 0) {
+    args.push("--model", input.model);
+  }
+  return args;
 }
 
 // src/lib/claude-code/job.ts

--- a/.github/actions/ferry-cc-prepare/cc-prepare-action.js
+++ b/.github/actions/ferry-cc-prepare/cc-prepare-action.js
@@ -6658,35 +6658,6 @@ function assertToolPolicyEnforcesNoAutoMerge(policy) {
   }
 }
 
-// src/lib/claude-code/claude-args.ts
-function buildClaudeArgs(input) {
-  const servers = input.mcpServers ?? [];
-  const allowedTools = [...nativeToolsForRole(input.role), ...mcpToolAllowlist(servers)];
-  const disallowedTools = [...NO_AUTO_MERGE_DENY];
-  assertToolPolicyEnforcesNoAutoMerge({ allowedTools, disallowedTools });
-  const args = [
-    "--append-system-prompt",
-    input.system,
-    "--allowedTools",
-    allowedTools.join(","),
-    "--disallowedTools",
-    disallowedTools.join(",")
-  ];
-  if (servers.length > 0) {
-    args.push("--mcp-config", JSON.stringify(toClaudeCodeMcpConfig(servers)));
-  }
-  if (input.maxTurns !== void 0) {
-    if (!Number.isInteger(input.maxTurns) || input.maxTurns <= 0) {
-      throw new Error(`max-turns must be a positive integer, got ${input.maxTurns}`);
-    }
-    args.push("--max-turns", String(input.maxTurns));
-  }
-  if (input.model !== void 0 && input.model.trim().length > 0) {
-    args.push("--model", input.model);
-  }
-  return args;
-}
-
 // src/lib/claude-code/output-artifact.ts
 var CC_OUTPUT_ARTIFACT_PATH = ".ferry/cc-output.json";
 var DONE_OUTCOMES = ["implemented", "already_satisfied", "blocked"];
@@ -6842,6 +6813,39 @@ FINAL OUTPUT (claude-code path): the \`done\`/\`finish_review\` tools are not av
   }
   const access = ROLE_ACCESS[role];
   return `${header} Commit and push your work with \`git\` first (there is no \`commit_progress\` tool on this path; access=${access}). Then write: ${DEV_ITER_SHAPE}`;
+}
+
+// src/lib/claude-code/claude-args.ts
+function buildClaudeArgs(input) {
+  const servers = input.mcpServers ?? [];
+  const allowedTools = [
+    ...nativeToolsForRole(input.role),
+    `Write(${CC_OUTPUT_ARTIFACT_PATH})`,
+    ...mcpToolAllowlist(servers)
+  ];
+  const disallowedTools = [...NO_AUTO_MERGE_DENY];
+  assertToolPolicyEnforcesNoAutoMerge({ allowedTools, disallowedTools });
+  const args = [
+    "--append-system-prompt",
+    input.system,
+    "--allowedTools",
+    allowedTools.join(","),
+    "--disallowedTools",
+    disallowedTools.join(",")
+  ];
+  if (servers.length > 0) {
+    args.push("--mcp-config", JSON.stringify(toClaudeCodeMcpConfig(servers)));
+  }
+  if (input.maxTurns !== void 0) {
+    if (!Number.isInteger(input.maxTurns) || input.maxTurns <= 0) {
+      throw new Error(`max-turns must be a positive integer, got ${input.maxTurns}`);
+    }
+    args.push("--max-turns", String(input.maxTurns));
+  }
+  if (input.model !== void 0 && input.model.trim().length > 0) {
+    args.push("--model", input.model);
+  }
+  return args;
 }
 
 // src/lib/claude-code/job.ts

--- a/src/lib/claude-code/claude-args.test.ts
+++ b/src/lib/claude-code/claude-args.test.ts
@@ -18,14 +18,24 @@ describe('buildClaudeArgs', () => {
 
   it('builds --allowedTools from the role native set (read-write for developer)', () => {
     const args = buildClaudeArgs({ role: 'developer', system: 's' });
-    expect(flagValue(args, '--allowedTools')).toBe('Bash,Read,Write,Edit,Glob,Grep');
+    expect(flagValue(args, '--allowedTools')).toBe(
+      'Bash,Read,Write,Edit,Glob,Grep,Write(.ferry/cc-output.json)',
+    );
   });
 
-  it('builds the read-only native set for reviewer and refiner', () => {
+  it('builds the read-only native set for reviewer and refiner with narrow artifact Write grant', () => {
     for (const role of ['reviewer', 'refiner'] as const) {
       expect(flagValue(buildClaudeArgs({ role, system: 's' }), '--allowedTools')).toBe(
-        'Read,Glob,Grep',
+        'Read,Glob,Grep,Write(.ferry/cc-output.json)',
       );
+    }
+  });
+
+  it('narrow Write(.ferry/cc-output.json) grant is present for every role', () => {
+    const ROLES: FerryRole[] = ['refiner', 'developer', 'reviewer', 'iterator'];
+    for (const role of ROLES) {
+      const allowed = flagValue(buildClaudeArgs({ role, system: 's' }), '--allowedTools') ?? '';
+      expect(allowed).toContain('Write(.ferry/cc-output.json)');
     }
   });
 
@@ -36,7 +46,7 @@ describe('buildClaudeArgs', () => {
     ];
     const args = buildClaudeArgs({ role: 'reviewer', system: 's', mcpServers: servers });
     expect(flagValue(args, '--allowedTools')).toBe(
-      'Read,Glob,Grep,mcp__jira__get_issue,mcp__context7',
+      'Read,Glob,Grep,Write(.ferry/cc-output.json),mcp__jira__get_issue,mcp__context7',
     );
   });
 

--- a/src/lib/claude-code/claude-args.ts
+++ b/src/lib/claude-code/claude-args.ts
@@ -18,6 +18,7 @@ import type { FerryRole } from './tool-profiles.js';
 import { nativeToolsForRole } from './tool-profiles.js';
 import { toClaudeCodeMcpConfig, mcpToolAllowlist } from './mcp-config.js';
 import { assertToolPolicyEnforcesNoAutoMerge, NO_AUTO_MERGE_DENY } from './tool-policy.js';
+import { CC_OUTPUT_ARTIFACT_PATH } from './output-artifact.js';
 
 export interface BuildClaudeArgsInput {
   role: FerryRole;
@@ -32,7 +33,14 @@ export interface BuildClaudeArgsInput {
 
 export function buildClaudeArgs(input: BuildClaudeArgsInput): string[] {
   const servers = input.mcpServers ?? [];
-  const allowedTools = [...nativeToolsForRole(input.role), ...mcpToolAllowlist(servers)];
+  // All roles need a narrow Write grant for the output artifact (.ferry/cc-output.json).
+  // Read-only roles (refiner/reviewer) have no broad Write in their native tool set,
+  // so without this grant the agent cannot write its result and cc-apply fails with ENOENT.
+  const allowedTools = [
+    ...nativeToolsForRole(input.role),
+    `Write(${CC_OUTPUT_ARTIFACT_PATH})`,
+    ...mcpToolAllowlist(servers),
+  ];
 
   // Fail-closed: assert the no-auto-merge invariant before emitting args.
   // Throws if the allow set re-grants a denied rule (ADR-0002 §D, #303).

--- a/src/lib/dispatch/cc-prepare-action.test.ts
+++ b/src/lib/dispatch/cc-prepare-action.test.ts
@@ -515,6 +515,10 @@ describe('runCcPrepareAction — non-refiner roles are wired into the entrypoint
       process.chdir(tmp.cwd);
       // GITHUB_TOKEN intentionally omitted — the role branch demands it via
       // `resolveRoleRuntimeContext`, which proves we reached the new code path.
+      // Explicitly clear GITHUB_TOKEN because the CI environment sets it, which
+      // would otherwise cause requireEnv('GITHUB_TOKEN') to succeed and reach
+      // runner.getRepoDefaultBranch before the expected throw.
+      vi.stubEnv('GITHUB_TOKEN', '');
       for (const [k, v] of Object.entries(
         baseEnv({ GITHUB_OUTPUT: tmp.outputFile, FERRY_AGENT_ROLE: role }),
       )) {

